src/corehost/cli/ijwhost/ijwhost.cpp

```diff
@@ -0,0 +1,152 @@
+#define IJW_API extern "C"
+#else
+#define IJW_API SHARED_API
+#endif // _WIN32
+
+pal::hresult_t get_load_in_memory_assembly_delegate(pal::dll_t handle, load_in_memory_assembly_fn* delegate)
+{
+    pal::dll_t fxr;
+
+    pal::string_t host_path;
+    if (!pal::get_own_module_path(&host_path) || !pal::realpath(&host_path))
+    {
+        trace::error(_X("Failed to resolve full path of the current host module [%s]"), host_path.c_str());
+        return StatusCode::CoreHostCurHostFindFailure;
+    }
+
+    pal::string_t dotnet_root;
+    pal::string_t fxr_path;
+    if (!resolve_fxr_path(get_directory(host_path), &dotnet_root, &fxr_path))
+    {
+        return StatusCode::CoreHostLibMissingFailure;
+    }
+
+    // Load library
+    if (!pal::load_library(&fxr_path, &fxr))
+    {
+        trace::error(_X("The library %s was found, but loading it from %s failed"), LIBFXR_NAME, fxr_path.c_str());
+        trace::error(_X("  - Installing .NET Core prerequisites might help resolve this problem."));
+        trace::error(_X("     %s"), DOTNET_CORE_INSTALL_PREREQUISITES_URL);
+        return StatusCode::CoreHostLibLoadFailure;
+    }
+
+    // Leak fxr
+
+    auto get_delegate_from_hostfxr = (hostfxr_get_delegate_fn)pal::get_symbol(fxr, "hostfxr_get_runtime_delegate");
+    if (get_delegate_from_hostfxr == nullptr)
+        return StatusCode::CoreHostEntryPointFailure;
+
+    pal::string_t app_path;
+
+    if (!pal::get_module_path(handle, &app_path))
+    {
+        trace::error(_X("Failed to resolve full path of the current mixed-mode module [%s]"), host_path.c_str());
+        return StatusCode::LibHostCurExeFindFailure;
+    }
+    
+    return get_delegate_from_hostfxr(host_path.c_str(), dotnet_root.c_str(), app_path.c_str(), hostfxr_delegate_type::load_in_memory_assembly, (void**)delegate);
+}
+
+IJW_API BOOL STDMETHODCALLTYPE _CorDllMain(HINSTANCE hInst,
+    DWORD  dwReason,
+    LPVOID lpReserved
+)
+{
+    BOOL res = TRUE;
+
+    PEDecoder pe(hInst);
+
+    // In the following code, want to make sure that we do our own initialization before
+    // we call into managed or unmanaged initialization code, and that we perform
+    // uninitialization after we call into managed or unmanaged uninitialization code.
+    // Thus, we do DLL_PROCESS_ATTACH work first, and DLL_PROCESS_DETACH work last.
+    if (dwReason == DLL_PROCESS_ATTACH)
+    {
+        // If this is not a .NET module (has a CorHeader), shouldn't be calling _CorDllMain
+        if (!pe.HasCorHeader())
+        {
+            return FALSE;
+        }
+
+        if (pe.IsILOnly() || (!pe.HasManagedEntryPoint() && !pe.HasNativeEntryPoint()))
+        {
+            // If there is no user entry point, then we don't want the
+            // thread start/stop events going through because it slows down
+            // thread creation operations
+            DisableThreadLibraryCalls(hInst);
+        }
+
+        // Install the bootstrap thunks
+        if (!patch_vtable_entries(pe))
```

Comment:
Another problem is `trace::error`. By default this writes to `stderr` even if tracing is disabled - so it does that in 99.9% of cases.

I don't know enough about the IJW scenarios to be able to tell what should be the right behavior. That said, by default the process will get failures into its `stderr` output, which might not be desirable.

In case of COM we redirect error writing to a buffer and report it through COM error reporting. The main scenario there is if something tries to load a COM component which fails, it should be able to handle the failure "Silently".

I don't know if IJW has such scenario where failure to load IJW assembly should also be possible to handle "Silently". But I think it's probably safer to do that. The one scenario which comes to mind is a simple `Assembly.LoadFile` on IJW assembly - which may fail and the caller would expect to be able to simply `try/catch-ignore` around it without any consequences. If we write errors to stderr that would not be the case.

Note that even if we redirect error writing, if tracing is enabled all error output goes to the trace as well - so that would be a way to debug such failures (enable host tracing to get full information).

Reply:
So there's two areas in the IJW host that trace errors/warnings:

1. During load, if we cannot change the page permissions on the memory page that the vtfixup table is located in, we log an error when it's unrecoverable and a warning when it's recoverable. In the .NETFX implementation, these logs are just debug-only auto-fail asserts instead, so that is another option for those.

2. When starting the runtime upon the first call to the native function. Since we're in an arbitrary call stack (like I mentioned above with the exception throw) we have no knowledge whatsoever of who is calling us or what scenario we are in. Additionally, this code-path is not executed at all if IJW is first loaded as a managed assembly since the .NET assembly loader patches the vtable with the JIT stubs before anything can call these thunks. I can't think of any good options for alternative error logging here nor can I think of any scenarios where we would want this to be handled "Silently".

Reply:
In both cases the errors seem recoverable from the outside caller perspective. Basically the logic is as follows. Let's say I have code like this:
``` C#
try
{
  Assembly wpfAssembly = Assembly.LoadFile("wpf_ijw_Assembly.dll");
}
catch (Exception ex)
{
  Console.WriteLine("Something failed, but we can live without WPF for now.");
}
```

If we want/need to support the above scenario then I would also expect the above code to not write anything to `stderr`. If we don't "swallow" error writing in the host, it will pollute the `stderr`. 

I understand that there's no good way to report detailed error information and I think it's OKish in this case. Debugging can be done by turning on host tracing (`COREHOST_TRACE=1`) which would include the errors as well any other tracing we do in the host.

But we simply should not write to `stderr` unless we're explicitly asked to in "component" scenarios. This is unfortunate consequence of the fact that host was original written as the owner of the process, in which case writing to `stderr` made sense. Since we've added lot more scenarios where the host doesn't own the process (comhost, ijwhost, nethost) or where `stderr` is a bad choice (GUI apps on Windows).

Reply:
In your example the second error cannot occur, in which case I agree with you. I'm totally fine "swallowing" the trace in the initial loader code.

Here's an example of the second case I came up with:

```c
void DoStuff(int(*callback)(int) );

int main()
{
     HMODULE mod = LoadLibrary(W("my-ijw-assembly.dll")); // In the first case, this is the call that will fail.
     int(*myFunc)(int) = GetProcAddress(mod, "MyFunc");
     DoStuff(myFunc);
}

void DoStuff(int(*callback)(int))
{
      // Lots of interesting code here
      // More interesting code

      callback(5); // In the second case the exception would be thrown here if this is the first call into managed code in the IJW assembly.
}
```

Since we're more likely than not going to take down the process with an unhandled exception in this case and also the fact that we're basically inserting all of the runtime startup code into an arbitrary stack somewhere in the user's process, I think silently failing here would not be good since unlike the COM case, there's no good way for us to notify the user that the runtime initialization failed in this case without logging.